src/main/java/org/opensearch/geospatial/ip2geo/cache/GeoDataCache.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.geospatial.ip2geo.cache;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+
+import org.opensearch.common.cache.Cache;
+import org.opensearch.common.cache.CacheBuilder;
+
+/**
+ * Cache to hold geo data
+ *
+ * GeoData in an index in immutable. Therefore, invalidation is not needed.
+ */
+public class GeoDataCache {
+    private Cache<CacheKey, Map<String, Object>> cache;
+
+    public GeoDataCache(final long maxSize) {
+        if (maxSize < 0) {
+            throw new IllegalArgumentException("ip2geo max cache size must be 0 or greater");
+        }
+        this.cache = CacheBuilder.<CacheKey, Map<String, Object>>builder().setMaximumWeight(maxSize).build();
+    }
+
+    public Map<String, Object> putIfAbsent(
+        final String indexName,
+        final String ip,
+        final Function<String, Map<String, Object>> retrieveFunction
+    ) throws ExecutionException {
+        CacheKey cacheKey = new CacheKey(indexName, ip);
+        return cache.computeIfAbsent(cacheKey, key -> retrieveFunction.apply(key.ip));
+    }
+
+    public Map<String, Object> get(final String indexName, final String ip) {
+        return cache.get(new CacheKey(indexName, ip));
+    }
+
+    /**
+     * Create a new cache with give size and replace existing cache
+     *
+     * Try to populate the existing value from previous cache to the new cache in best effort
+     *
+     * @param maxSize
+     */
+    public void updateMaxSize(final long maxSize) {
+        if (maxSize < 0) {
+            throw new IllegalArgumentException("ip2geo max cache size must be 0 or greater");
+        }
+        Cache<CacheKey, Map<String, Object>> temp = CacheBuilder.<CacheKey, Map<String, Object>>builder().setMaximumWeight(maxSize).build();
+        int count = 0;
+        Iterator<GeoDataCache.CacheKey> it = cache.keys().iterator();
+        while (it.hasNext() && count < maxSize) {
```

Comment:
This is an interesting condition. We are providing the update to max size, and if new size if less than the old size we are copying less data. This adds more doubt in my mind why we are even providing this functionality.

Reply:
This is not to throw away all the cached value when size is changed.